File: python/Binary Tree Postorder Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    answer = []
    def rec(self, root):
        if not root:
            return
        else:
            self.rec(root.left)
            self.rec(root.right)
            self.answer.append(root.val)
    
    def postorderTraversal(self, root):
        self.answer = []
        self.rec(root)
        return self.answer

File: cpp/Binary Tree Postorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        if (root == NULL) {
            return vector<int>();
        }
        vector<int> answer;
        stack<TreeNode*> st;
        
        while (!st.empty() || root != NULL) {
            if (root) {
                st.push(root);
                answer.push_back(root->val);
                root = root->right;
                continue;
            } else {
                root = st.top()->left;
                st.pop();
            }
        }
        
        reverse(answer.begin(), answer.end());
        return answer;
    }
    
    
    vector<int> answer;
    vector<int> postorderTraversal_rec(TreeNode* root) {
        helper(root);
        return answer;
    }
    
    void helper(TreeNode *root) {
        if (root == NULL) {
            return;
        }
        helper(root->left);
        helper(root->right);
        answer.push_back(root->val);
    }
};
